core/src/test/java/org/elasticsearch/common/xcontent/XContentParserTests.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import org.elasticsearch.test.ESTestCase;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasSize;
+
+public class XContentParserTests extends ESTestCase {
+
+    public void testReadList() throws IOException {
+        Map<String, Integer> sources = new HashMap<>();
+        // Normal usage
+        sources.put("{\"foo\": [\"bar\"]}", 1);
+        sources.put("{\"foo\": [\"bar\",\"baz\"]}", 2);
+        sources.put("{\"foo\": [{\"bar\":1},{\"baz\":2},{\"qux\":3}]}", 3);
+        sources.put("{\"foo\": [1]}", 1);
+        sources.put("{\"foo\": [1,2]}", 2);
+        // Less common but possible usage
+        sources.put("{\"foo\": \"bar\"}", 1);
+        sources.put("{\"foo\": 1}", 1);
+        sources.put("{\"foo\": {\"bar\":\"baz\"}}", 1);
+        sources.put("{\"foo\": [{},{},{},{}]}", 4);
```

Comment:
Can you test with eg. `{"foo": 1, "bar": 2}`? I'm afraid calling parser.list() on foo might return [1,2] given how it's implemented?